Userland/Utilities/telws.cpp

```diff
@@ -0,0 +1,105 @@
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <AK/Format.h>
+#include <AK/URL.h>
+#include <LibCore/ArgsParser.h>
+#include <LibCore/EventLoop.h>
+#include <LibCore/File.h>
+#include <LibCore/Notifier.h>
+#include <LibWebSocket/ConnectionInfo.h>
+#include <LibWebSocket/Message.h>
+#include <LibWebSocket/WebSocket.h>
+
+int main(int argc, char** argv)
+{
+    Core::ArgsParser args_parser;
+
+    const char* origin = nullptr;
+    String url_string;
+
+    args_parser.add_positional_argument(url_string, "URL to connect to", "url", Core::ArgsParser::Required::Yes);
+    args_parser.add_option(origin, "URL to use as origin", "origin", 'o', "origin");
+
+    args_parser.parse(argc, argv);
+
+    URL url(url_string);
+
+    if (!url.is_valid()) {
+        warnln("The given URL is not valid");
+        return 1;
+    }
+
+    WebSocket::ConnectionInfo connection_info(url);
+    connection_info.set_origin(String(origin));
+
+    Core::EventLoop loop;
+
+    auto socket = WebSocket::WebSocket::create(connection_info);
+    socket->on_open = [&]() {
+        outln("[WebSocket opened]");
+    };
+    socket->on_error = [](auto error) {
+        outln("[WebSocket Error : {}]", (unsigned)error);
+    };
+    socket->on_message = [](auto message) {
+        if (!message.is_text()) {
+            outln("[Received binary data : {} bytes]", message.data().size());
+            return;
+        }
+        outln("[Received utf8 text] {}", String(ReadonlyBytes(message.data())));
+    };
+    socket->on_close = [&](auto code, auto message, bool was_clean) {
+        outln("[Server {} closed connection : {} (code {})]",
+            was_clean ? "cleanly" : "dirtily",
+            message,
+            code);
+        loop.quit(0);
+    };
+    socket->start();
+
+    auto stdin = Core::File::standard_input();
+    auto stdin_watcher = Core::Notifier::construct(stdin->fd(), Core::Notifier::Read);
+    stdin_watcher->on_ready_to_read = [&] {
+        if (stdin->can_read_line()) {
+            auto line = stdin->read_line();
```

Comment:
Since this is line-based, LibLine can probably give you a better interface, though it might mess with the event loop too much.
(this is coming from "a barebones REPL", trying to get rid of the "barebones" lol)

Reply:
I looked into it, and I think `Line::Editor::get_line` is blocking, right ? That might not mesh well with the event loop, since the server will regularily ping the client to maintain the connection.
In any wase, integrating telws with LibLine seems like a big task, and a bit out of scope. The `telws` program is mostly temporary anyways, since most of the work on WebSockets should happen on the JS side. So it's hopefully not too bad if it's barebones !

Reply:
it will keep the event loop going, LibLine itself is event-based : )
The integration should just be 
```c++
auto editor = Line::Editor::construct();
...
auto line_or_error = editor->get_line();
```
if that doesn't work, then nvm.

Reply:
Ok, it "works", but libline reallly doesn't like when stuff outputs to the terminal while it's waiting for a prompt. Added it anyways :)